### PR TITLE
Add GPU CI: e2e tests and benchmarks

### PR DESCRIPTION
**Pull request by ryan-williams**
## Summary

- Add deterministic e2e training test (`tests/e2e_train.py`) with platform-specific expected values
- Add GPU e2e workflow (`gpu-e2e.yml`) using [ec2-gha] on EC2 `g6.xlarge` (NVIDIA L4), runs on PRs targeting `main`
- Add GPU benchmark workflow (`gpu-benchmark.yml`) with configurable model size, weekly schedule, and manual dispatch
- Add WandB logging to benchmark: logs training metrics, model config, dataset version, and instance type
- Add `scripts/s3_sync.py`: reusable S3 data sync with size filtering and deterministic dataset hashing
- Add `gen-expected.yml` for regenerating expected values on GHA runners (macOS, Ubuntu)
- Add `e2e_training_demo.ipynb` notebook with training visualization
- Add `.github/workflows/README.md` documenting all workflows
- Rename `tests/electrai/` → `tests/test_electrai/` to fix import shadowing
- Add `--gradient-checkpoint` flag to `e2e_train.py` for large models on limited VRAM
- Benchmark summary includes linked commit SHA for traceability

## Passing Runs

- [GPU E2E Test #27](https://github.com/Quantum-Accelerators/electrai/actions/runs/22358546405) — GPU + CPU baseline, both pass (Feb 24)
- [GPU Benchmark #27](https://github.com/Quantum-Accelerators/electrai/actions/runs/22358551091) — 50 S3 samples, WandB logging (Feb 24)

## Required Setup

Secrets:
- `GH_SA_TOKEN` — GitHub PAT for runner registration
- `WANDB_API_KEY` — WandB API key (optional, for benchmark logging; currently Ryan's personal key)

IAM/OIDC:
- Trust policy configured in [Open-Athena/ops] for `ec2-gha` OIDC authentication

## Screenshots

<table>
<tr>
<td align="center">
  <strong>EC2 instances</strong>
  <br>
  <a href="https://gist.githubusercontent.com/ryan-williams/dbd94c3f5669f88f946c99e1d37818f6/raw/deb69187d25cc22e8451ff3f9bbfe2ad75690e33/instances.png">
    
  </a>
</td>
<td align="center">
  <strong>EC2 tags</strong>
  <br>
  <a href="https://gist.githubusercontent.com/ryan-williams/dbd94c3f5669f88f946c99e1d37818f6/raw/deb69187d25cc22e8451ff3f9bbfe2ad75690e33/tags.png">
    
  </a>
</td>
</tr>
<tr>
<td align="center" colspan="2">
  <strong>
    <a href="https://github.com/Quantum-Accelerators/electrai/actions/runners?tab=self-hosted">
      GitHub runners
    </a>
  </strong>
  <br>
  <a href="https://gist.githubusercontent.com/ryan-williams/dbd94c3f5669f88f946c99e1d37818f6/raw/deb69187d25cc22e8451ff3f9bbfe2ad75690e33/runners.png">
    
  </a>
</td>
</tr>
<tr>
<td align="center" colspan="2">
  <strong>
    <a href="https://wandb.ai/PrinceOA/elf-net-ci">
      WandB dashboard
    </a>
  </strong>
  <br>
  <a href="https://gist.githubusercontent.com/ryan-williams/dbd94c3f5669f88f946c99e1d37818f6/raw/e8e71904f7094f76dbfbf0a9c91349e3099db9a4/bench.gif">
    
  </a>
</td>
</tr>
</table>

## Test plan
- [x] GPU e2e test passes on EC2 `g6.xlarge`
- [x] CPU e2e test passes on EC2
- [x] GPU benchmark completes with production-size model (32ch/16 blocks, 128^3 grids)
- [x] `gen-expected.yml` generates correct values on macOS and Ubuntu
- [x] Expected values verified across all 3 platforms (`darwin-arm64`, `linux`, `linux-gpu`)
- [x] WandB run logged with correct metadata (project, dataset version, instance type)
- [x] `scripts/s3_sync.py` downloads correct samples and generates deterministic dataset hash

[ec2-gha]: https://github.com/Open-Athena/ec2-gha
[Open-Athena/ops]: https://github.com/Open-Athena/ops
[Quantum-Accelerators/electrai#64]: https://github.com/Quantum-Accelerators/electrai/pull/64

### CODE REVIEW

File: .github/workflows/README.md [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+**Inputs:**
+- `instance_type`: EC2 instance type (default: `g6.xlarge`)
+- `epochs`: Training epochs (default: `5`)
+- `channels`: Model channels - 8=tiny, 32=prod, 64=large (default: `32`)
+- `residual_blocks`: Residual blocks - 2=tiny, 16=prod (default: `16`)
+- `s3_samples`: Number of S3 samples to download (0=use repo test data, default: `0`)
+- `debug`: Debug mode (default: `false`)
+
+**Note:** By default uses the 5-sample repo test data (`s3_samples=0`). Set `s3_samples=10` or higher to download samples from S3 for more realistic benchmarks.
+
```
[the remark this answers is not in the document]
Comment (ryan-williams) [reply .github/workflows/README.md]: Updated the `gpu-benchmark.yml` section with current defaults (`s3_samples=50`), weekly schedule, WandB logging, all inputs, and added `WANDB_API_KEY` to the secrets section. Fixed in ebd8b3d.

**Comment by ryan-williams** [comment]
Fixed `test_e2e_train.py` to use `--epochs 5` matching `expected_values.json`. Also removed dead `LossTracker` class, moved `import time` to top of function body, and parameterized WandB entity via `WANDB_ENTITY` env var. Fixed in 4ab3c10.

File: tests/e2e_train.py [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+    if wandb_project and logger.experiment:
+        logger.experiment.summary["wallclock_s"] = train_wallclock
+        run_url = logger.experiment.get_url()
+        if run_url:
+            echo(f"WANDB_RUN_URL={run_url}")
+
+    # Get final losses
+    final_val_loss = trainer.callback_metrics.get("val_loss_epoch")
+    if final_val_loss is None:
+        final_val_loss = trainer.callback_metrics.get("val_loss")
```
[the remark this answers is not in the document]
Comment (ryan-williams) [reply tests/e2e_train.py]: Added a guard: if both `val_loss_epoch` and `val_loss` are missing, prints a clear error and exits. Fixed in e37fd22.

File: scripts/s3_sync.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+
+# fmt: off
+@click.command()
+@click.option("-b", "--bucket", default="openathena", help="S3 bucket name")
+@click.option("-M", "--max-file-size", default=25.0, type=float, help="Skip files larger than N MB (0=no limit)")
+@click.option("-n", "--num-samples", default=10, type=int, help="Max samples to download")
+@click.option("-o", "--output", default="data/s3", help="Output directory")
+@click.option("-p", "--prefix", default="electrai", help="S3 key prefix")
+@click.option("-v", "--verbose", is_flag=True, help="Verbose output")
```
[the remark this answers is not in the document]
Comment (ryan-williams) [reply scripts/s3_sync.py]: Replaced the `# fmt: off` block with a targeted `# fmt: skip` on the one long line. Ruff's `RUF028` doesn't allow `# fmt: on` between decorators and function defs, so this is cleaner. Fixed in 264f256.

File: tests/e2e_train.py [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+    class Config:
+        pass
+
+    cfg = Config()
+
+    # Dataset
+    cfg.dataset_name = "mp"
+    cfg.data_path = data_path or str(repo_root / "data/MP/chgcars/input")
+    cfg.label_path = label_path or str(repo_root / "data/MP/chgcars/label")
+    cfg.map_path = map_path or str(repo_root / "data/MP/map/map_sample.json.gz")
```
Comment (forklady42) [inline tests/e2e_train.py]: Is this supposed to match our current config? We shifted patterns since this PR started 😅. Now relying on [a txt file](https://github.com/Quantum-Accelerators/electrai/blob/edc327156bbd71896cee5eae31524bb1fd058246/src/electrai/configs/MP/config_resunet.yaml#L4) similar to the `map_path` with `data` and `label` directories under the same root path.

**Review by forklady42 — APPROVED** [review]
Claude gave me several additional comments, but I got it to admit that most of the suggestions were wrong. 😂 

I've also taken a look. At this point, I think it's worth merging and seeing how the CI works in practice. We can address any issues as they arise.
